Add i2c dump filtering

barnas-michal opened the pull request:
This PR changes format of the i2c messages dump, to align the messages for easier comparison, and in case of device emulators, it's using the target device name instead of generic "emul".
It also adds a possibility to filter which devices transactions should be dumped. If someone is debugging the i2c communication, the log being flooded with other devices' transactions make it hard to debug. The logged devices
should be specified in the device-tree using the compatible string.

Example of device-tree node:
    i2c-dump-filter {
            compatible = "zephyr,i2c-dump-filter";
            devices = < &display0 >, < &sensor3 >;
    };

barnas-michal committed:
i2c: use device instead of name for i2c dump messages

This commit changes the parameter of i2c_dump_msgs function from
string name to pointer to the device structure.
It allows for comparison of device pointers and allow to use
the printed device name in i2c shell commands.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
i2c: improve the message dumps messages

This commit changes the format of printed messages to align the
following strings and make it more readable.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
i2c: add filtering of i2c dumped messages

This commit adds option to dump i2c messages of only specified
devices. It makes it easier to debug communication of specific
i2c device instead of logging all i2c communication.
The filter of devices is specifiec in device-tree using the
node with "zephyr,i2c-dump-filter" compatible string.

Example of device-tree node:
i2c-dump-filter {
	compatible = "zephyr,i2c-dump-filter";
	devices = < &display0 >, < &sensor3 >;
};

Signed-off-by: Michał Barnaś <mb@semihalf.com>

galak reviewed (dismissed):
I'm not fan of this as a devicetree node.  Maybe a property under a controller or something.

semihalf-barnas-michal commented:
> I'm not fan of this as a devicetree node. Maybe a property under a controller or something.

I am not sure if I would be able to easily iterate over all devices with some property set. It would also require to add the optional property to all possible i2c devices.
I know that this node doesn't represent any physical device, but I think it is the easiest and most readable approach to this functionality.

tmon-nordic commented:
> I know that this node doesn't represent any physical device, but I think it is the easiest and most readable approach to this functionality.

I agree that this is the easiest approach. However, there is constantly a disagreement (not only in Zephyr, but also in Linux kernel) what constitutes a devicetree and what does not. Linux kernel tends to be stricter about not using devicetree for configuration purposes.

Unfortunately I cannot find my older comments about the same issue, but this is perfect example of a place where some "devicetree-like-configuration" would make sense. Maybe we should group such use cases into some issue?

teburd reviewed (dismissed):
I like the way this works.

I was under the impression that zephyr compatibles were understood to be zephyr/virtual-ish like devices and configuration.

I don't see how this would easily be done in Kconfig. It wants structured data as configuration and DT is the tool Zephyr has for that.

aaronemassey reviewed (dismissed):
@semihalf-barnas-michal 
> I know that this node doesn't represent any physical device, but I think it is the easiest and most readable approach to this functionality.

@tmon-nordic 
> I agree that this is the easiest approach. 

In my opinion, this is really similar to the first acknowledged type of **exception** scenario listed in Zephyr's docs [comparing DeviceTree with KConfig](https://docs.zephyrproject.org/latest/build/dts/dt-vs-kconfig.html). So it's acceptable to me as devicetree mechanism.

keith-zephyr reviewed (dismissed):
Non-blocking nits from me.

keith-zephyr commented:
> > I'm not fan of this as a devicetree node. Maybe a property under a controller or something.
> 
> I am not sure if I would be able to easily iterate over all devices with some property set. It would also require to add the optional property to all possible i2c devices. I know that this node doesn't represent any physical device, but I think it is the easiest and most readable approach to this functionality.

The GPIO hogs driver provides an example for iterating over all devicetree nodes that have the `gpio-controller` property. Something similar could be done here by adding properties to the i2c-controller.yaml base class.  But considering this is a debug feature I like the single devicetree node that collects up the list of I2C devices to log.

semihalf-barnas-michal commented on drivers/i2c/Kconfig:
```diff
@@ -30,12 +30,25 @@ config I2C_STATS
+	depends on I2C_DUMP_MESSAGES
+	help
+	  Filter dumped messages for specified buses and addresses.
+	  The filter table is defined in the device-tree using the compatible string of
```
Done

semihalf-barnas-michal commented on drivers/i2c/Kconfig:
```diff
@@ -30,12 +30,25 @@ config I2C_STATS
+	  Filter dumped messages for specified buses and addresses.
+	  The filter table is defined in the device-tree using the compatible string of
+	  "zephyr,i2c-dump-filter" and phandles to the devices that need to be traced.
+	  Example of device-tree node:
```
Done

semihalf-barnas-michal commented:
> hogs driver provides an example for iterating over all devicetree nodes

I've taken a look at it. The difference is that gpio controllers include this yaml, however the i2c filter should be having a reference to device on the i2c bus, not the controller. It probably may have a child-binding, but then there could be a problem with devices that are specified using multiple children due to usage of more than one address on the bus.

semihalf-barnas-michal commented:
Is there anything more needed to do before merge?

teburd commented:
@galak If you could take another look please that would be helpful here

teburd commented:
@semihalf-barnas-michal Perhaps it would be beneficial to document this debug feature in the i2c bus docs somewhere, as an opt in feature to dump messages, or dump messages with filtering abilities using a dts node.

At the moment someone would need to read the source code to understand how to use this feature.

barnas-michal committed:
doc: fix missing and remove obsolete links

Fix text that should be linking to the URLs but was missing
the link reference and remove the obsolete links.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
doc: extract the documentation about debugging to another file

Move the chapters about Application Debugging and
Debugging using Eclipse to another file for easier search and
to allow more debugging chapters to be inserted here.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
doc: add documenatation about I2C_DUMP_MESSAGES

Add the documentation about I2C_DUMP_MESSAGES and about the
I2C_DUMP_MESSAGES_FILTER in the develop/debug/index.rst doc file.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

semihalf-barnas-michal commented:
> in the i2c bus docs somewhere

I've extracted some of the debugging documentation to another file and added the documentation about this feature there. If you do not like it, I can revert and just add the documentation to the i2c doc file.

teburd commented:
> > in the i2c bus docs somewhere
> 
> I've extracted some of the debugging documentation to another file and added the documentation about this feature there. If you do not like it, I can revert and just add the documentation to the i2c doc file. I see that there is another file, services/debug which may be in a conflict to my extraction... However I am not sure if the I2C documentation is suitable for this, since it is very long and mostly describing the I2C interface and API...

Great thank you, I think as long as there's some documentation somewhere on it, it will be helpful. If it turns out to be in the wrong place that's easily fixed.

keith-zephyr commented on doc/develop/debug/index.rst:
```diff
@@ -286,6 +286,81 @@ Support for Zephyr RTOS awareness is implemented in `pyOCD v0.11.0`_ and later.
+Debugging I2C communication
+***************************
+
+There is a possibility to log all or some of the I2C transactions done by application.
```
```suggestion
There is a possibility to log all or some of the I2C transactions done by the application.
```

keith-zephyr commented on doc/develop/debug/index.rst:
```diff
@@ -286,6 +286,81 @@ Support for Zephyr RTOS awareness is implemented in `pyOCD v0.11.0`_ and later.
+
+There is a possibility to log all or some of the I2C transactions done by application.
+This feature is enabled by the Kconfig option :kconfig:option:`CONFIG_I2C_DUMP_MESSAGES`, but it
+uses the ``LOG_DBG`` function to print the contents so also the I2C logging level has to be set
+to the debug level. It can be done by enabling the :kconfig:option:`CONFIG_I2C_LOG_LEVEL_DBG` option.
```
```suggestion
uses the ``LOG_DBG`` function to print the contents so the  :kconfig:option:`CONFIG_I2C_LOG_LEVEL_DBG` option must also be enabled.
```

keith-zephyr commented on doc/develop/debug/index.rst:
```diff
@@ -286,6 +286,81 @@ Support for Zephyr RTOS awareness is implemented in `pyOCD v0.11.0`_ and later.
+   D: contents:
+   D: 43 42 41 00 00 00 00 00 |CBA.....
+
+The first line shows that the log belongs to the I2C transactions and informs which I2C interface
+was used to sent it and what is the address of the device targeted.
```
```suggestion
The first line indicates the I2C controller and the target address of the transaction.
```

keith-zephyr commented on doc/develop/debug/index.rst:
```diff
@@ -286,6 +286,81 @@ Support for Zephyr RTOS awareness is implemented in `pyOCD v0.11.0`_ and later.
+
+The first line shows that the log belongs to the I2C transactions and informs which I2C interface
+was used to sent it and what is the address of the device targeted.
+In above example, the I2C interface is named ``io_i2c_ctrl7_port0`` and the target device address
```
```suggestion
In above example, the I2C controller is named ``io_i2c_ctrl7_port0`` and the target device address
```

keith-zephyr commented on doc/develop/debug/index.rst:
```diff
@@ -286,6 +286,81 @@ Support for Zephyr RTOS awareness is implemented in `pyOCD v0.11.0`_ and later.
+
+   the address, length and contents values are in hexadecimal, but lack the ``0x`` prefix
+
+Next lines contain messages, both sent and received, and content of read messages being dumped
+depending on the parameter of the function ``i2c_dump_msgs_rw``. This function is available for use
```
```suggestion
Next lines contain messages, both sent and received.  The contents of write messages is always shown, while the content of read messages is controlled by a parameter to the function ``i2c_dump_msgs_rw``. This function is available for use

```

keith-zephyr commented on doc/develop/debug/index.rst:
```diff
@@ -286,6 +286,81 @@ Support for Zephyr RTOS awareness is implemented in `pyOCD v0.11.0`_ and later.
+
+Next lines contain messages, both sent and received, and content of read messages being dumped
+depending on the parameter of the function ``i2c_dump_msgs_rw``. This function is available for use
+by user, but is also called internally by ``i2c_transfer`` API function with content dump enabled.
```
```suggestion
by user, but is also called internally by ``i2c_transfer`` API function with read content dump enabled.
```

keith-zephyr commented on doc/develop/debug/index.rst:
```diff
@@ -286,6 +286,81 @@ Support for Zephyr RTOS awareness is implemented in `pyOCD v0.11.0`_ and later.
+Filtering the I2C communication dump
+====================================
+
+During the development, it may be needed to investigate communication with one or more target
+devices. By default, all I2C communication is logged with every device on the bus.
+It may litter the log with unrelated devices and make it impossible to effectively debug the
+communication with device of interest.
+To resolve this issue, there is a feature that allows to select with which devices the communication
+should be logged.
+
+To enable this feature a Kconfig option :kconfig:option:`CONFIG_I2C_DUMP_MESSAGES_FILTER`
+must be enabled. The devices are selected using the devicetree node, for example::
```
```suggestion
By default, all I2C communication is logged between all I2C controllers and I2C targets.
It may litter the log with unrelated devices and make it difficult to effectively debug the
communication with a device of interest.

Enable the Kconfig option :kconfig:option:`CONFIG_I2C_DUMP_MESSAGES_FILTER` to create an allowlist of I2C targets to log. The allowlist of devices is configured using the devicetree, for example::
```

keith-zephyr commented on doc/develop/debug/index.rst:
```diff
@@ -286,6 +286,81 @@ Support for Zephyr RTOS awareness is implemented in `pyOCD v0.11.0`_ and later.
+      };
+
+      i2c-dump-filter {
+          compatible = "zephyr,i2c-dump-filter";
```
After reviewing the documentation, would it be clearer to call this option `CONFIG_I2C_DUMP_MESSAGES_ALLOWLIST` and the compatible `zephyr,i2c-dump-allowlist`?

semihalf-barnas-michal replied:
Yes, it will be a better name for this feature. I will change that and update this PR.

teburd commented on doc/develop/application/index.rst:
```diff
@@ -431,7 +431,7 @@ should know about.
 * :makevar:`SHIELD`: see :ref:`shields`
 
-* :makevar:`ZEPHYR_MODULES`: A CMake list containing absolute paths of
+* :makevar:`ZEPHYR_MODULES`: A `CMake list`_ containing absolute paths of
```
This seems like a stray change? Approved already but noticed after

kartben replied:
it's an external link

https://github.com/zephyrproject-rtos/zephyr/blob/d670f1ba6ecd7988c4220025513f55895c40f6c4/doc/develop/application/index.rst?plain=1#L1291

https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#external-links

teburd commented on doc/develop/modules.rst:
```diff
@@ -276,7 +276,7 @@ The merging of pull requests in the main branch of a module
 file update in the zephyr main tree.
 
-**Issue Reporting:** GitHub issues are intentionally disabled in module
+**Issue Reporting:** `GitHub issues`_ are intentionally disabled in module
```
why is this being single quoted

teburd reviewed (changes requested):
Some of the single quoted additions to the docs don't make much sense to me at a quick glance
